### PR TITLE
ARTEMIS-2420 Adding support for DLA/DLQ prefix for wildcard addresses

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

**Comment by clebertsuconic** [comment]
@PiotrKlimczak Please review this

**Comment by PiotrKlimczak** [comment]
@clebertsuconic I don't have permission to push to this branch.
Should I create separate PR with more changes?

**Comment by clebertsuconic** [comment]
Do you have more changes? You won't have access to my branch. This one is mine.


if you want to add more changes, do it on your fork:
Do the following script:

```sh
git remote add clebert https://github.com/clebertsuconic/activemq-artemis.git
git fetch clebert
git checkout clebert/ARTEMIS-2420 -B ARTEMIS-2420
git push origin ARTEMIS-2420
```

Do your stuff:


```sh
git commit -a --amend
git push origin -f
```


send the Pull Request.

i will close this when you have done it.

**Comment by clebertsuconic** [comment]
Please, do not use merge commits on Pull Requests... do a full rebase with 

git pull rebase upstream master
then push -f

**Comment by clebertsuconic** [comment]
The following test is failing on your changes:

org.apache.activemq.artemis.tests.integration.management.QueueControlTest.testGetDeadLetterAddress

**Comment by PiotrKlimczak** [comment]
I think It might be good to create PR to your PR branch, so you will see what the differences are.
Regarding the failing test, which profiles have you run with maven?

**Comment by clebertsuconic** [comment]
we run with -Ptests, but if you run this on Idea  (or the IDE of your preference) you should see it.

**Comment by clebertsuconic** [comment]
I will wait for your changes.

**Comment by michaelpearce-gain** [comment]
A few bits. What occurs if both dead letter address is set and prefix? I assume it is an either or, as such should enforce this in xsd and also check when configuring.

What occurs if queue/address is allowed to be created due security settings, but the user able to do that is unable to create the dead letter address? Do both fail? Does one part succeed but the dla fails? And then what occurs?

Is there auto clean up, so the auto created dead letter addresses are removed when the original address is removed, in cases of auto creation.

**Comment by michaelpearce-gain** [comment]
For the mentioned cases tests also probably needed

**Comment by PiotrKlimczak** [comment]
OK so I am running all tests now so taking this time to write some design documentation of what was implemented how and why.

Let's start with example configuration:

         <address-setting match="#">
            <dead-letter-address-auto-create prefix="DLA.">
               <routing-type>AS_ORIGIN</routing-type>
               <durable>true</durable>
               <temporary>false</temporary>
            </dead-letter-address-auto-create>
         </address-setting>
         <address-setting match="DLA.#">
             
         </address-setting>

So how DLAs will behave, can be configured in standard way.
Also there is new routing type added for DLA addresses only: AS_ORIGIN.
This new type ensures that if you have multiple queues for address, the message will go to corresponding DLQ only. So if you have address A1 with queues Q1 and Q2, and Q2 consumption fails, the message will go only to DLA.Q2 DLQ and never to DLQ.Q1.
This ensures predictive dead letter outcome if configured, as obviously ANYCAST and MULTICAST (default) are also supported.

**Comment by michaelpearce-gain** [comment]
Added some inline comments also

**Comment by michaelpearce-gain** [comment]
I think @PiotrKlimczak configuration will more flexible. Same failure edge cases and constraints i commented above would still apply and need to be resolved though. Also personally any defaults e.g. route type, temporary should probably take from the parent queue, e.g. you wouldnt want an anycast message routing to a multicast dead letter address and vice versa by default not multicast as mentioned.

**Comment by michaelpearce-gain** [comment]
Another concern or thing to think about, is what if auto creation is disabled in address settings match that the auto created queue was configured to go into, should this actually be checked at point of configuration to avoid such miss configurations?

**Comment by michaelpearce-gain** [comment]
Also would it be possible to configure a spefiic queue to goto a specific address, taking precedence over the default prefix when set.

**Comment by michaelpearce-gain** [comment]
@PiotrKlimczak just confirmation i like the concept. I do personally prefer the configurability that you propose over @clebertsuconic impl in this pr.

That said because this isnt a simple small feature mostly as we can span address and security setting matches that complicates things especially in the miss configuration, and other cases that need to be factored in. 

It may take several iterations to iron them all out.

**Comment by PiotrKlimczak** [comment]
@michaelpearce-gain Both what is in this pr and what I proposed came from me.
The thing is that I have started reading about Artemis few weeks ago and only few days ago I've seen its code 1st time.
So I suppose it is improving through me getting more knowledge with your help guys.
Thanks for that a lot and for your patience to me :).

Now I have created PR to this branch, so you can see the changes which I am working on.
https://github.com/clebertsuconic/activemq-artemis/pull/29
Not fully finished yet as not all concerns are addressed and requires more test coverage.
But basic test coverage proves it works correctly.
Also it is not fully building corredtly due to some rat licensing issue, which I need to solve.

**Comment by michaelpearce-gain** [comment]
Idea around security and address creation restrictions, could that both the address and queue settings are checked both for the new address/queue and the dead letter, before it is created. Thus meaning that either both create (because permission and settings allow) or both wouldnt. Thus avoiding the situation of one being created without the other

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd [hunk 1]
```diff
@@ -2939,6 +2939,13 @@
                   </xsd:documentation>
                </xsd:annotation>
             </xsd:element>
+            <xsd:element maxOccurs="1" minOccurs="0" name="dead-letter-address-prefix" type="xsd:string">
```
[the remark this answers is not in the document]
Comment (clebertsuconic) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: Can you do that on the XSD? or should be done when parsing?

File: artemis-server/src/test/resources/ConfigurationTest-full-config.xml [hunk 2]
```diff
@@ -377,6 +377,9 @@
       </security-settings>
 
       <address-settings>
+         <address-setting match="#">
+            <dead-letter-address-prefix>DLA.</dead-letter-address-prefix>
```
[the remark this answers is not in the document]
Comment (clebertsuconic) [reply artemis-server/src/test/resources/ConfigurationTest-full-config.xml]: Hmmmm... for actually the security settings on this case.. shouldn't it be the same as the queue?

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelpearce-gain) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: Yes. Its xsd choice.

Also we should have similar check if we have other ways of configuring address settings, not 100% if we do. But cli and jmx as possible points come to mind

File: artemis-server/src/test/resources/ConfigurationTest-full-config.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelpearce-gain) [reply artemis-server/src/test/resources/ConfigurationTest-full-config.xml]: Why should it? As suggested in the case of new auto dla then maybe both settings are checked before a queue is created only if both are possible then you could craete queue

**Comment by clebertsuconic** [comment]
@michaelandrepearce I was thinking the same. and was going to suggest it.


It would be better if Piotr opened his own pr. I was only opening this PR to cleanup the previous one.

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd (hunk 1 above)
[the remark this answers is not in the document]
Comment (PiotrKlimczak) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: You cannot have choice in all. I don't think you can achieve this without breaking backwards compatibility.
Will add that to requirements however.

[the remark this answers is not in the document]
Comment (michaelpearce-gain) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: So all was recent change. It used to sequence i think we should be able to change it back to sequence

**Comment by michaelpearce-gain** [comment]
Also to add. Once all sorted for dla/dlq. The same should be made for expiryQueues with all the same rules and logic.

Little point having dynamic dla/dlq but not for expiries.

**Comment by PiotrKlimczak** [comment]
I have summarised all questions and answers from this PR on https://jira.apache.org/jira/browse/ARTEMIS-2420.
I think this can be closed now to avoid missing comments.

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd (hunk 1 above)
[the remark this answers is not in the document]
Comment (PiotrKlimczak) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: I have checked git history, unfortunately, it goes back to 2014.
I can change it to sequence, as long as we are happy to accept regression caused and expectation from users to change their configs.
My personal feeling is that it is not worth it, as instead, I can implement that in java, giving the same effect but not affecting users.
I am happy however to follow your guidance if you think that is best for the project.
Please let me know.

**Comment by PiotrKlimczak** [comment]
> Also to add. Once all sorted for dla/dlq. The same should be made for expiryQueues with all the same rules and logic.

> Little point having dynamic dla/dlq but not for expiries.

@michaelpearce-gain 
Can you please confirm that you want to do expiration under the same PR as DLQs?

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelpearce-gain) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: Java is fine.

**Comment by clebertsuconic** [comment]
@PiotrKlimczak are you sending your own PR? I want to close this one as I only opened to cleanup your PR.

**Comment by PiotrKlimczak** [comment]
@clebertsuconic Sorry I was on holidays.
I am not creating PR yet a @michaelpearce-gain requested to complete all the tasks before it is considered for merge as he wants all of them in one go instead of phases.
I have consolidated all requirements under https://jira.apache.org/jira/browse/ARTEMIS-2420 and progress too.

**Comment by clebertsuconic** [comment]
@PiotrKlimczak ok, I wish close this PR then.

I wish we could move all the discussion to a new PR.


I suggest you open a discuss thread on the dev list for any further discussions instead of keeping a technical discussion on JIRA or github.

Small discussions around the source code are great here, but for longer discussions we prefer to keep them on the dev-list so it's recorded.

**Comment by PiotrKlimczak** [comment]
@clebertsuconic 
Can you please suggest a dev list? More precisely can you please paste a link?

I have expiry addresses auto-creation now working too.
So I am getting closer to finish.
However, I have problems with configurability as using all instead of sequence in XSD is killing lots of features.
I have an idea how to migrate to all to sequences which will not affect users.
This will, however, require bit time to implement.
Simply I don't like the fact that you can configure parameters for DLA auto-creation, which doesn't make sense and due to all instead of sequence in XSD, validations capabilities are very limited.

Thanks.

**Comment by rcsilva83** [comment]
Please, support suffix on "dead-letter-address-auto-create" element too. We prefer "Queue.Name.DLQ" so DLQ queues stay close to the original ones in alphabetical order. This is how our current ActiveMQ is configured.

**Comment by peterstiles** [comment]
Is there a similar capability for "Expiry"?

**Pull request closed without merging**